tech_notes/tn008_auto_insert_cells.md

```diff
@@ -0,0 +1,395 @@
+might do the following
+
+1. Use `kubectl get deploy` to get the flux annotation  `kustomize.toolkit.fluxcd.io/name` to identify the kustomization
+2. Use `kubectl get kustomization` to get the last applied revision of the customization and to identify the source
+   controller
+3. Use `kubectl get gitrepository` to get the source repository and its latest commit
+
+In this case the command at each step depends on the output of the previous step.
+
+
+### Troubleshooting Workload Identity and GCP Permissions
+
+Troubleshooting workflows are often multi-step workflows. For example, suppose we are trying to troubleshoot why a
+pod on a GKE cluster can't access a GCS bucket. We might do the following
+
+1. Use kubectl to get the deployment to identify the Kubernetes service account (KSA)
+2. Use kubectl to get the KSA to identify and get the annotation specifying the GCP service account (GSA)
+3. Use `gcloud policy-troubleshoot iam` to check whether the GSA has the necessary permissions to access the GCS bucket
+
+## Motivation: Verbose Responses and Chat Models
+
+One of the problems with the existing UX is that since we use Chat models to generate completions the completions
+often contain markdown cells before or after the code cells.  These cells correspond to the chat model providing additional
+exposition or context. For example if we use a cell with the text
+
+```
+Use gcloud to list the buckets
+```
+
+Foyle inserts two cells
+
+A markdown cell
+
+```
+To list the buckets using `gcloud`, you can use the following command:
+```
+
+and then a code cell
+
+```bash
+gcloud storage buckets list
+```
+
+The markdown cell is redundant with the previous markdown cell and a user would typically delete it. We'd like
+to create a better UX where the user can more easily accept the suggested code cell and reject the markdown cell.
+
+
+## UX For Continuous Suggestion
+
+As the user edits one cell (either markdown or code), we'd like Foyle to be continuously running in the background to 
+generate one or more cells to be inserted after the current cell. This is similar to GitHub Copilot and 
+[Continue.Dev Autocomplete](https://docs.continue.dev/walkthroughs/tab-autocomplete. An important difference
+is that we don't need Foyle to autocomplete the current cell. This should simplify the problem because 
+* we have more time to generate completions
+  * the completion needs to be ready by the time the user is ready to move onto the next cell rather than before
+    they type the next character
+* we can use larger models and longer context windows since we lave a larger latency budget
+* we don't need to check that the completion is still valid after each character is typed into the current cell because we aren't
+trying to auto complete the word(s) and current cell.
+
+If we mimicked ghost text. The experience might look as follows
+
+* User is editing cell N in the notebook
```

Comment:
Conceptually, I believe this will work. However, since building on the VS Code Extension platform is a challenge of working inside of an opinionated platform like iOS/Android, I'd suggest proceeding with one of two different approaches or a combination of both.

1. Prototype the notebook UX entirely using fake/fixture data to prove the concept.
2. Use the `StatusProvider` API to expose the "Rejection" and/or "Completion" buttons inside cells to work out the exact interaction ahead of using notebook/cell events.

Experience has taught me that these APIs aren't as flexible and versatile as native browser/DOM APIs, and working out the details can take a lot of time.